src/tools.js

```diff
@@ -1,341 +0,0 @@
-    return {
-        type: matches[1],
-        username: index >= 0 ? auth.substr(0, index) : auth,
-        password: index >= 0 ? auth.substr(index + 1) : '',
-    };
-};
-
-module.exports.parseProxyAuthorizationHeader = parseProxyAuthorizationHeader;
-
-/**
- * Works like Bash tee, but instead of passing output to file,
- * passes output to log
- *
- * @param   {String}   name          identifier
- * @param   {Boolean}  initialOnly   log only initial chunk of data
- * @return  {through}                duplex stream (pipe)
-
-export const tee = (name, initialOnly = true) => {
-    console.log('tee');
-    let maxChunks = 2;
-    const duplex = through((chunk) => {
-        if (maxChunks || !initialOnly) {
-            // let msg = chunk.toString();
-            // msg += '';
-            maxChunks--;
-            console.log(`pipe: ${JSON.stringify({
-                context: name,
-                chunkHead: chunk.toString().slice(0, 100),
-            })}`);
-        }
-        duplex.queue(chunk);
-    });
-
-    return duplex;
-};
-*/
-
-const addHeader = (headers, name, value) => {
-    if (headers[name] === undefined) {
-        headers[name] = value;
-    } else if (Array.isArray(headers[name])) {
-        headers[name].push(value);
-    } else {
-        headers[name] = [
-            headers[name],
-            value,
-        ];
-    }
-};
-
-module.exports.addHeader = addHeader;
-
-const PORT_SELECTION_CONFIG = {
-    FROM: 20000,
-    TO: 60000,
-    RETRY_COUNT: 10,
-};
-
-module.exports.PORT_SELECTION_CONFIG = PORT_SELECTION_CONFIG;
-
-const maybeAddProxyAuthorizationHeader = (parsedUrl, headers) => {
-    if (parsedUrl && (parsedUrl.username || parsedUrl.password)) {
-        // According to RFC 7617 (see https://tools.ietf.org/html/rfc7617#page-5):
-        //  "Furthermore, a user-id containing a colon character is invalid, as
-        //   the first colon in a user-pass string separates user-id and password
-        //   from one another; text after the first colon is part of the password.
-        //   User-ids containing colons cannot be encoded in user-pass strings."
-        // So to be correct and avoid strange errors later, we just throw an error
-        if (/:/.test(parsedUrl.username)) throw new Error('The proxy username cannot contain the colon (:) character according to RFC 7617.');
-        const auth = `${parsedUrl.username || ''}:${parsedUrl.password || ''}`;
-        headers['Proxy-Authorization'] = `Basic ${Buffer.from(auth).toString('base64')}`;
-    }
-};
-
-module.exports.maybeAddProxyAuthorizationHeader = maybeAddProxyAuthorizationHeader;
-
-// Replacement for Bluebird's Promise.nodeify()
-const nodeify = (promise, callback) => {
-    if (typeof callback !== 'function') return promise;
-
-    const p = promise.then((result) => callback(null, result), callback);
-
-    // Handle error from callback function
-    p.catch((e) => {
-        setTimeout(() => { throw e; }, 0);
-    });
```

Comment:
Not sure about removing this. Without it, the code produces unhandled rejections. With it, it produces uncaught exceptions. It's semantics, but they are handled differently. Since the function is `nodeify` and should produce a callback interface, it should not produce rejections when the callback throws because that's not how a callback interface behaves. You also should not be able to catch the error downstream by adding a catch handler to the promise later.

On the other hand, I have no idea how important it is for the package, it just caught my eye.

Reply:
Hmmm... You're right. This doesn't crash the process on Node.js 14, but does on 16. I'll revert this change.

Reply:
I was initially thinking about removing this from `proxy-chain`, but this would be too breaking as it's used in too many places. Maybe someday in the future.